design/unnumbered-bgp.md

```diff
@@ -0,0 +1,436 @@
+
+The observation is that peering is being established even if RA are not enabled on both side.
+Docs are unclear if needed.
+
+#### BGP RouterID
+
+//TODO is there a requirement for that?
+
+### Traffic forwarding
+
+FRR advertises IPv4 prefixes with IPv6 next hops using encoding specified in RFC 5549 (now RFC 8950).
+
+### The API Extension
+
+#### frr-k8s side
+
+```go
+...
+
+type Neighbor struct {
+  Description string `json:"description"` // optional
+  Address string `json:"address"` // must be empty if interface != nil
+  Interfaces []string `json:"interface"` //  
+  PeerGroup string `json:"peerGroup"` // optional, otherwise value <interface>-peergroup
+  RemoteAS string `json:"remoteAS"` // it can be "ASN number or internal or external [or auto]"
+  Port uint16 `json:"port,omitempty"`
+  PasswordSecret v1.SecretReference `json:"password,omitempty"`
+  HoldTime metav1.Duration `json:"holdTime,omitempty"`
+  KeepaliveTime metav1.Duration `json:"keepaliveTime,omitempty"`
+  EBGPMultiHop bool `json:"ebgpMultiHop,omitempty"`
+  BFDProfile string `json:"bfdProfile,omitempty"`
+  ToAdvertise Advertise `json:"toAdvertise,omitempty"`
+  ToReceive Receive `json:"toReceive,omitempty"`
+}
+
+...
+```
+
+#### MetalLB
+
+```go
+...
+type BGPPeerSpec struct {
+	MyASN uint32 `json:"myASN"`
+	ASN uint32 `json:"peerASN"` // or internal or external
+	Address string `json:"peerAddress"` // or interface name e.g. net0
+	SrcAddress string `json:"sourceAddress,omitempty"`
+	Port uint16 `json:"peerPort,omitempty"`
+	HoldTime metav1.Duration `json:"holdTime,omitempty"`
+	KeepaliveTime metav1.Duration `json:"keepaliveTime,omitempty"`
+	ConnectTime *metav1.Duration `json:"connectTime,omitempty"`
+	RouterID string `json:"routerID,omitempty"`
+	NodeSelectors []metav1.LabelSelector `json:"nodeSelectors,omitempty"`
+	Password string `json:"password,omitempty"`
+	PasswordSecret v1.SecretReference `json:"passwordSecret,omitempty"`
+	BFDProfile string `json:"bfdProfile,omitempty"`
+	EnableGracefulRestart bool `json:"enableGracefulRestart,omitempty"`
+	EBGPMultiHop bool `json:"ebgpMultiHop,omitempty"`
+	VRFName string `json:"vrf,omitempty"`
+	DisableMP bool `json:"disableMP,omitempty"`
+}
+
+```
+
+Example CR
+
+```yaml
+apiVersion: metallb.io/v1beta2
+kind: BGPPeer
+metadata:
+  name: peer1
+  namespace: metallb-system
+spec:
+  myASN: 64512
+  peerASN: external
+  peerAddress: net0
+```
+
+The resulting configuration will be something like:
+
+```yaml
+bgp:
+  routers:
+    - asn: 64512
+      neighbors:
+      //TODO
+```
+
+
+### Testing
+
+#### E2E tests
+
+- Modify the testing infra to include point-to-point links which means
+connect containers using VETH, and not through a bridge. We need NET-ADMIN cap (or sudo) when creating the VETH
+- Create/Delete peer per specific test not to take resources from runners outside the runtime of the test
+- This peer needs different configuration (not IPs, interface names), keep a raw config
```

Comment:
iiuc this will all be covered in the initial implementation?

Reply:
Can you elaborate what you mean?

Reply:
asking if these points will be covered or there's still a question about them, while the last point (about primary interfaces) still needs to be thought of